### DISCUSSION TITLE
Using PYOPTSPARSE_IMPORT_SNOPT_FROM

### DISCUSSION THREAD

**Discussion by swryan** [mdolab/pyoptsparse#396] [Installation] [answered]
This is not strictly a `pyoptsparse `issue, but I was trying to exercise the capability offered by the `PYOPTSPARSE_IMPORT_SNOPT_FROM` environment variable.

I seem to have successfully built the `SNOPT` library per the instructions [here](https://ccom.ucsd.edu/~optimizers/docs/snopt/install.html), and created `snopt*.so` using the provided `f2py` file,  but have not been able to close the loop with `pyoptsparse`.

As can be seen [here](https://github.com/swryan/pyoptsparse/actions/runs/8971636926) on GitHub, I get errors when trying to import `snopt` as built.

I was hoping someone with some experience using `SNOPT` in this way with `pyoptsparse` could point me in the right direction.

Here is the relevant fragment of my script:
```
            echo "============================================================="
            echo "Building SNOPT"
            echo "============================================================="
            cd SNOPT
            ./configure
            make
            make check

            cd lib
            if [[ "${{ matrix.OS }}" == "ubuntu-latest" ]]; then
              cp /home/runner/work/pyoptsparse/pyoptsparse/pyoptsparse/pySNOPT/source/f2py/snopt.pyf .
            else
              cp /Users/runner/work/pyoptsparse/pyoptsparse/pyoptsparse/pySNOPT/source/f2py/snopt.pyf .
            fi
            python -m numpy.f2py -c -m snopt snopt.pyf

            echo "============================================================="
            echo "Try to import SNOPT"
            echo "============================================================="
            python -c "import snopt"
```
And here are the results..

Ubuntu/Conda:
```
=============================================================
Try to import SNOPT
=============================================================
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: /home/runner/work/pyoptsparse/pyoptsparse/SNOPT/lib/snopt.cpython-312-x86_64-linux-gnu.so: undefined symbol: openunit_
```
MacOS/Conda:
```
=============================================================
Try to import SNOPT
=============================================================
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: dlopen(/Users/runner/work/pyoptsparse/pyoptsparse/SNOPT/lib/snopt.cpython-312-darwin.so, 0x0002): symbol not found in flat namespace '_sqlog_'
```

**Comment by ewu63** [comment, marked as answer]
pySNOPT requires a few more things to be linked during compilation (for example the error points to [`openunit`](https://github.com/mdolab/pyoptsparse/blob/main/pyoptsparse/pySNOPT/source/openunit.f) not found. [Here](https://github.com/mdolab/pyoptsparse/issues/335#issuecomment-1690100023) is a thread with more clear instructions (which we will try to add to our docs...), in particular note the line
```
SRC=pyoptsparse/pyoptsparse/pySNOPT
python -m numpy.f2py -m snopt -c ${SRC}/f2py/snopt.pyf ${SRC}/*.f -L${PREFIX}/lib -lsnopt7
```

The `${SRC}/*.f` argument tells f2py to additionally build bindings for these fortran files that are shipped with pyOptSparse.

**Reply by swryan** [reply]
Thanks again.. I was able to get there...  this is what I ended up with:
```
            cd SNOPT
            ./configure
            make
            make check

            if [[ "${{ matrix.OS }}" == "ubuntu-latest" ]]; then
              PYSNOPT_SOURCE=/home/runner/work/pyoptsparse/pyoptsparse/pyoptsparse/pySNOPT/source/
            else
              PYSNOPT_SOURCE=/Users/runner/work/pyoptsparse/pyoptsparse/pyoptsparse/pySNOPT/source/
            fi

            cp -v $PYSNOPT_SOURCE/*.f src/
            cp -v $PYSNOPT_SOURCE/f2py/snopt.pyf .

            rm src/snopth.f
            rm src/sn27lu77.f
            rm src/sn27lu90.f
            
            python -m numpy.f2py -m snopt -c snopt.pyf src/*.f -L`pwd`/lib -lsnopt7

            python -c "import snopt; print(snopt)"

            export PYOPTSPARSE_IMPORT_SNOPT_FROM=`pwd`
```

**Reply by swryan** [reply]
Here is a test workflow, in case it's useful:

https://github.com/swryan/pyoptsparse/blob/test_snopt/.github/workflows/test-snopt.yml

**Comment by swryan** [comment]
ah great..  let me try that, thanks so much 👍🏼
